### PR TITLE
SONARHTML-137 Add new rule detection S8697

### PR DESCRIPTION
**Pull request by erwan-leforestier-sonarsource**
## Summary                                                                    
  Implements rule [S8697](https://github.com/SonarSource/rspec/pull/XXXX)       
  (*"srcset" entries should have a width or pixel density descriptor*). Resolves
   [SONARHTML-137](https://jira.sonarsource.com/browse/SONARHTML-137).          
                                                                                
  The new `SrcSetDescriptorCheck` flags `` and `<source srcset>`    
  entries that omit (or use an invalid) descriptor. Multi-entry `srcset` values
  must explicitly declare a width (`400w`) or pixel density (`1x`, `2x`, `.5x`) 
  descriptor on every entry.                                                                               
                                                                      
  ## Changes                                                                    
  - New check: `SrcSetDescriptorCheck.java`                                     
  - Unit tests: `SrcSetDescriptorCheckTest.java` + `valid-cases.html` +         
  `invalid-cases.html`                                                          
                                                                                
  ## FV                                                                  
  - [ ] `mvn test -pl sonar-html-plugin` passes                                 
  - [ ] `mvn verify -Pits` ruling tests pass with the new expected file         

[SONARHTML-137]: https://sonarsource.atlassian.net/browse/SONARHTML-137?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: sonar-html-plugin/src/main/java/org/sonar/plugins/html/checks/sonar/SrcSetDescriptorCheck.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    if (!this.isTargetedNode(node)) {
+      return;
+    }
+
+    // Read the value in srcset
+    String srcSetValue = node.getAttribute("srcset");
+
+    if (srcSetValue == null) {
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (erwan-leforestier-sonarsource) [reply sonar-html-plugin/src/main/java/org/sonar/plugins/html/checks/sonar/SrcSetDescriptorCheck.java:56]: Will be fixed

File: sonar-html-plugin/src/main/java/org/sonar/plugins/html/checks/sonar/SrcSetDescriptorCheck.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    
+    if (Helpers.isDynamicValue(srcSetValue, getHtmlSourceCode())) {
+      return;
+    }
+
+    // Split the value by comma so we can extract the difference sources
+    String[] sources = srcSetValue.split(",");
+
+    if (sources.length < 2) {
+      return;
```
[the remark this answers is not in the document]
Comment (erwan-leforestier-sonarsource) [reply sonar-html-plugin/src/main/java/org/sonar/plugins/html/checks/sonar/SrcSetDescriptorCheck.java:62]: Will be fixed

**Review by vdiez — CHANGES_REQUESTED** [review]
Reviewed against SONARHTML-137 and the RSPEC PR.

There are still two parser issues I think we should fix before merge:

- `sonar-html-plugin/src/main/java/org/sonar/plugins/html/checks/sonar/SrcSetDescriptorCheck.java:49-50`: `srcSetValue.split(",")` is not a valid `srcset` parser. A valid candidate URL can itself contain commas, for example `data:image/svg+xml,%3Csvg%3E 1x, logo-2x.png 2x`. The current split turns that into bogus fragments and would raise a false positive.
- `sonar-html-plugin/src/main/java/org/sonar/plugins/html/checks/sonar/SrcSetDescriptorCheck.java:33-34` together with `:56-61`: the check only looks for a descriptor-shaped suffix via `matcher.find()`, so invalid candidates such as `logo.png 1x 2x` or `logo.png 400w 2x` are accepted as compliant. These should be rejected.

Could we tighten the parsing and add tests for both cases? I’m happy to re-review after that.

**Comment by erwan-leforestier-sonarsource** [comment]
Thanks for the review @vdiez 🙏 To address your comment, the two issues you found are relevant so I've modified the way I was parsing the srcset, by manually going through the srcset value to extract the urls and descriptors in the most efficient way. I've added tests as well to make sure the corner cases you mentioned are properly handled.
See my commit: https://github.com/SonarSource/sonar-html/pull/656/changes/39f82ac865e7ad294c73ee1a1eb732e4e0e04821

**Review by vdiez — CHANGES_REQUESTED** [review]
Re-reviewing `09d3105`.

The parser rewrite addresses the two issues from my previous review: we no longer split on commas blindly, and candidates with multiple descriptors are now rejected.

I still think we have one blocker before merge in `sonar-html-plugin/src/main/java/org/sonar/plugins/html/checks/sonar/SrcSetDescriptorCheck.java:35-36`. The current regex is stricter than the HTML `srcset` grammar in some places and looser in another:

- it rejects valid width descriptors such as `01w`
- it rejects valid density descriptors such as `1e1x`
- it accepts invalid density descriptors such as `1.x`

Per the HTML spec, a width descriptor is a valid non-negative integer greater than zero followed by `w`, and a density descriptor is a valid floating-point number greater than zero followed by `x`. In other words, the rule still has both false positives and false negatives if we want spec-validity here.

I re-checked the single-candidate `<source media|type>` carve-out as well. I am not blocking on that: if we intentionally want to exempt that pattern entirely, the current behavior is defensible. But in that case I would probably make the rule description a bit more explicit that the exemption also skips malformed explicit descriptors, not just omitted ones.

Smaller follow-up, not blocking for this PR: `srcset="logo.png ${density}, logo-2x.png 2x"` is currently reported because we only treat the candidate as dynamic when the URL token is dynamic, not when the descriptor token is.

**Comment by vdiez** [comment]
Small clarification on the non-blocking `media` / `type` note from my latest review: I?m not arguing that a single-candidate `<source>` with `media` or `type` must require a descriptor.

The narrower point is this: if `media` / `type` are present, but there is also a `srcset`, we should probably still validate that the `srcset` syntax is correct when the author explicitly writes a descriptor.

So treating this as exempt seems reasonable:
`<source media="(min-width: 800px)" srcset="hero.jpg">`

But treating these as exempt is a different choice:
`<source media="(min-width: 800px)" srcset="hero.jpg 80%">`
`<source type="image/webp" srcset="hero.jpg 1x 2x">`

If the intended policy is to exempt the whole single-candidate `<source media|type>` case, including malformed explicit descriptors, that is defensible too; I would just make that explicit in the rule description.

File: sonar-html-plugin/src/main/resources/org/sonar/l10n/web/rules/Web/S8697.html [hunk 3]
```diff
@@ -9,7 +9,9 @@ <h2>Why is this an issue?</h2>
 <p>When a descriptor is omitted or invalid, the entry falls back to <code>1x</code>. While allowed by the spec, it is good practice to define a valid
 and explicit descriptor on every entry to make the developer’s intent visible and avoid silent fallbacks.</p>
 <p>This rule does not flag a single-URL <code>srcset</code> on a <code>&lt;source&gt;</code> element when <code>media</code> or <code>type</code> is
-set: those are art direction and format negotiation patterns, where the parent attribute already discriminates.</p>
+set <strong>and no descriptor is written</strong>: those are art direction and format negotiation patterns, where the parent attribute already
```
Comment (vdiez) [inline sonar-html-plugin/src/main/resources/org/sonar/l10n/web/rules/Web/S8697.html:12]: you need to modify rspec, otherwise it will be overwritten next time we run rule-api

**Comment by erwan-leforestier-sonarsource** [comment]
Thanks @vdiez, makes sense — narrowed the carve-out so an explicit descriptor on `<source media|type>` is still validated (malformed ones like `srcset="hero.jpg 80%"` now flagged), and made the policy explicit in the rule description: [rspec/6a43b52d](https://github.com/SonarSource/rspec/pull/6840/changes/6a43b52dbe56da15ba8325764a7897fc7233b273).

**Comment by vdiez** [comment]
check [latest review](https://github.com/SonarSource/sonar-html/pull/656#pullrequestreview-4345360960). This is not yet handled:

>  The current regex is stricter than the HTML srcset grammar in some places and looser in another:
> 
> it rejects valid width descriptors such as 01w
> it rejects valid density descriptors such as 1e1x
> it accepts invalid density descriptors such as 1.x
> Per the HTML spec, a width descriptor is a valid non-negative integer greater than zero followed by w, and a density descriptor is a valid floating-point number greater than zero followed by x. In other words, the rule still has both false positives and false negatives if we want spec-validity here.

**Comment by erwan-leforestier-sonarsource** [comment]
Thanks @vdiez, all three fixed in [74bfc3eb](https://github.com/SonarSource/sonar-html/pull/656/commits/74bfc3eb):

- `01w` now accepted (width allows leading zeros, only the value must be > 0).
- `1e1x` now accepted (exponent allowed per the HTML floating-point grammar).
- `1.x` now rejected (a `.` requires fractional digits).

Locked in via new entries in `valid-cases.html` / `invalid-cases.html`.

For the non-blocking follow-up about dynamic descriptor tokens (`srcset="logo.png ${density}, …"`) — happy to bundle it here if you prefer, otherwise I will leave it for a separate ticket.

**Review by vdiez — APPROVED** [review]
Re-reviewed `dd0e94d`.

The remaining blockers are fixed now:

- the descriptor validation matches the HTML grammar for the cases we discussed (`01w`, `1e1x`, `1.x` / `1.e0x`)
- the single-candidate `<source media|type>` carve-out is now correctly limited to the no-descriptor case, with matching rule documentation and tests

I re-ran `mvn -pl sonar-html-plugin -Dtest=SrcSetDescriptorCheckTest test` on this head and it passes.

The mixed dynamic-descriptor case (`srcset="logo.png ${density}, ..."`) still looks like a possible follow-up false positive, but I agree that does not need to block this PR.

**Comment by erwan-leforestier-sonarsource** [comment]
Thanks @vdiez 🙏 merging now

**Comment by vdiez** [comment]
don't forget to merge your rspec PRs @erwan-leforestier-sonarsource
